examples/server/server.cpp

```diff
@@ -1136,28 +1136,19 @@ struct server_context {
         if (!system_prompt.empty()) {
             system_tokens = ::llama_tokenize(ctx, system_prompt, true);
 
-            llama_batch_clear(batch);
+            const int32_t n_batch = llama_n_batch(ctx);
+            const int32_t n_tokens_prompt = system_tokens.size();
 
-            for (int i = 0; i < (int)system_tokens.size(); ++i) {
-                llama_batch_add(batch, system_tokens[i], i, { 0 }, false);
-            }
+            for (int32_t i = 0; i < n_tokens_prompt; i += n_batch) {
+                const int32_t n_tokens = std::min(n_batch, n_tokens_prompt - i);
 
-            const int32_t n_batch = llama_n_batch(ctx);
+                llama_batch_clear(batch);
 
-            for (int32_t i = 0; i < batch.n_tokens; i += n_batch) {
```

Comment:
I guess there is some outdated logic in `update_slots` as well, because there is a similar loop there.

Reply:
The only problem I see in `update_slots` is when the batch size is smaller than the number of slots, in which case tokens are added to the batch anyway.

https://github.com/ggerganov/llama.cpp/blob/06943a69f678fb32829ff06d9c18367b17d4b361/examples/server/server.cpp#L2002

When the batch size is bigger than the number of slots, it does not seem to have the same problem as the system prompt, because there is already never more than `n_batch` tokens added to a batch:

https://github.com/ggerganov/llama.cpp/blob/06943a69f678fb32829ff06d9c18367b17d4b361/examples/server/server.cpp#L2243-L2254

So yes, this loop:

https://github.com/ggerganov/llama.cpp/blob/06943a69f678fb32829ff06d9c18367b17d4b361/examples/server/server.cpp#L2313

might be a bit outdated, but I think it's still relevant for the error handling logic:

https://github.com/ggerganov/llama.cpp/blob/06943a69f678fb32829ff06d9c18367b17d4b361/examples/server/server.cpp#L2375-L2376

From the above I think what could be fixed is parallel generation with small batch sizes:

```diff
diff --git a/examples/server/server.cpp b/examples/server/server.cpp
index a1bbafbc..b2bf5bb5 100644
--- a/examples/server/server.cpp
+++ b/examples/server/server.cpp
@@ -753,13 +753,13 @@ struct server_context {
         default_generation_settings_for_props = get_formated_generation(slots.front());
         default_generation_settings_for_props["seed"] = -1;
 
-        // the update_slots() logic will always submit a maximum of n_batch tokens
+        // the update_slots() logic will always submit a maximum of n_batch or n_parralel tokens
         // note that n_batch can be > n_ctx (e.g. for non-causal attention models such as BERT where the KV cache is not used)
         {
             const int32_t n_batch = llama_n_batch(ctx);
 
             // only a single seq_id per token is needed
-            batch = llama_batch_init(n_batch, 0, 1);
+            batch = llama_batch_init(std::max(n_batch, params.n_parallel), 0, 1);
         }
 
         metrics.init();
```